poc/test_opaque_ake.sage

```diff
@@ -180,103 +145,102 @@ def test_3DH():
-            outputs["registration_response"] = to_hex(reg_response.serialize())
-            outputs["registration_upload"] = to_hex(record.serialize())
-            outputs["KE1"] = to_hex(ke1)
-            outputs["KE2"] = to_hex(ke2)
-            outputs["KE3"] = to_hex(ke3)
-            outputs["session_key"] = to_hex(server_session_key)
-            outputs["export_key"] = to_hex(export_key)
-
-            vector = {}
-            vector["config"] = client_kex.json()
-            vector["inputs"] = inputs
-            vector["intermediates"] = intermediates
-            vector["outputs"] = outputs
-            vectors.append(vector)
+            for (idU, idS) in [(None, None), (idU, None), (None, idS), (idU, idS)]:
+                (skU, pkU) = group.key_gen()
+                (skS, pkS) = group.key_gen()
+                skU_bytes = group.serialize_scalar(skU)
+                pkU_bytes = group.serialize(pkU)
+                skS_bytes = group.serialize_scalar(skS)
+                pkS_bytes = group.serialize(pkS)
+                oprf_seed = random_bytes(fast_hash().digest_size)
+
+                kdf = HKDF(fast_hash)
+                mac = HMAC(fast_hash)
+                config = Configuration(mode, oprf, kdf, mac, fast_hash, mhf, group)
+
+                creds = Credentials(skU_bytes, pkU_bytes, idU, idS)
+                core = OPAQUECore(config)
+
+                reg_request, metadata = core.create_registration_request(pwdU)
+                reg_response, kU = core.create_registration_response(reg_request, pkS_bytes, oprf_seed, credential_identifier)
+                record, export_key = core.finalize_request(creds, pwdU, metadata, reg_response)
+
+                # TODO(caw): do something else with this
+                pkU_enc = record.pkU
+                pkU = group.deserialize(pkU_enc)
+
+                client_kex = OPAQUE3DH(config)
+                server_kex = OPAQUE3DH(config)
+
+                ke1 = client_kex.generate_ke1(pwdU, info1, idU, pkU, idS, pkS)
+                ke2 = server_kex.generate_ke2(ke1, oprf_seed, credential_identifier, record.envU, record.masking_key, info2, idS, skS, pkS, idU, pkU)
+                ke3 = client_kex.generate_ke3(ke2)
+                server_session_key = server_kex.finish(ke3)
+
+                assert server_session_key == client_kex.session_key
+
+                inputs = {}
+                intermediates = {}
+                outputs = {}
+
+                # Protocol inputs
+                if idU:
+                    inputs["client_identity"] = to_hex(idU)
+                if idS:
+                    inputs["server_identity"] = to_hex(idS)
+                inputs["oprf_seed"] = to_hex(oprf_seed)
+                inputs["credential_identifier"] = to_hex(credential_identifier)
+                inputs["password"] = to_hex(pwdU)            
+                inputs["client_private_key"] = to_hex(skU_bytes)
+                inputs["client_public_key"] = to_hex(pkU_bytes)
```

Comment:
One problem: when these are derived from the randomized_pwd seed, these are no longer inputs -- they are intermediate values. In particular, I don't think these were updated as their corresponding values when derived from the randomized_pwd seed, and so they end up not matching...

Reply:
I confirm